src/main/scala/org/constellation/rewards/RewardsManager.scala

```diff
@@ -0,0 +1,201 @@
+    snapshot.checkpointBlocks.toList
+      .traverse(checkpointService.fullData)
+      .map(_.flatten.flatMap(_.checkpointBlock.observations))
+
+  private def updateRewards(distribution: Map[String, Double]): F[Unit] =
+    rewards.modify(prevDistribution => (prevDistribution |+| distribution, ()))
+
+  /*
+    TODO: Not sure if it is the correct way of updating balances by rewards.
+    1. putUnsafe takes Account.hash but we pass there Id.address which could differ. Maybe we should do Address(id.address).hash.
+    2. I take the current balance and just copy the balance by adding rewards. Seems to be quite... dirty and unsafe?
+   */
+  private def updateAddressBalances(rewards: Map[String, Double]): F[Unit] =
+    rewards.transform {
+      case (address, reward) =>
+        addressService.lookup(address).flatMap { existingCacheData =>
+          val updatedAddressCacheData = existingCacheData.map(
+            cache =>
+              cache.copy(
+                balance = cache.balance + reward.toLong,
+                memPoolBalance = cache.memPoolBalance + reward.toLong
+              )
+          ).getOrElse(AddressCacheData(reward.toLong, reward.toLong))
+
+          addressService.putUnsafe(address, updatedAddressCacheData)
+        }
+    }.values.toList.sequence.void
+}
+
+object RewardsManager {
+  val roundingError = 0.000000000001
+
+  /*
+  Rewards computed about one snapshots per 3 minutes, i.e. about 14,600 snapshots per month.
+  Snapshots for a first epoch of 2.5 years.
+   */
+  val epochOne = 438000 // = 14,600 snapshots per month * 12 months per year * 2.5 years
+  val epochTwo = 876000 // = 2 * epochOne (5 years)
+  val epochThree = 1314000 // = 3 * epochOne (7.5 years)
+  val epochFour = 1752000 // = 4 * epochOne (10 years)
+  val rewardsPool = epochFour
+
+  /*
+  10,000 units per month.
+   */
+  val epochOneRewards = 0.68493150684 // = 10,000 / 14,600
+  val epochTwoRewards = 0.34246575342 // = epochOneRewards / 2
+  val epochThreeRewards = 0.17123287671 // = epochTwoRewards / 2
+  val epochFourRewards = 0.08561643835 // = epochThreeRewards / 2
+
+  /*
+  Partitioning of address space, light nodes have smaller basis that full.
+  Normalizes rewards based on node size.
+   */
+  val partitonChart = Map[String, Set[String]]()
+
+  /*
+  Should come from reputation service.
+   */
+  val transitiveReputationMatrix = Map[String, Map[String, Double]]()
+  val neighborhoodReputationMatrix = Map[String, Double]()
+
+  def validatorRewards(
+    curShapshot: Int,
+    transitiveReputationMatrix: Map[String, Map[String, Double]],
+    neighborhoodReputationMatrix: Map[String, Double],
+    partitonChart: Map[String, Set[String]]
+  ) = {
+    val trustEntropyMap = shannonEntropy(transitiveReputationMatrix, neighborhoodReputationMatrix)
+    val distro = rewardDistribution(partitonChart, trustEntropyMap)
+    distro.mapValues(_ * rewardDuringEpoch(curShapshot))
+  }
```

Comment:
@buckysballs I haven't used that part 🤔 Not sure if that part is just a leftover or I actually have some GAPs in my way of thinking about rewards

Reply:
Those are from the older implementation. The transitiveReputationMatrix is used to weight values for light nodes. rewardDistribution looks good here